### ISSUE TITLE
RFC: Spinnaker Plugin System

### ISSUE THREAD

**Issue by andrewbackes** [spinnaker/spinnaker#4181]
# RFC: Spinnaker Plugin System

A plugin system for Spinnaker would be a way for users to add functionality without having to rebuild and maintain the entire project or go through pull requests to extend the project for their own purposes. This means that functionality can be added at boot/runtime. The code for a plugin also does not have to live in the core Spinnaker codebase.

### Extensibility

Enterprises have a rich set of internal tooling. Spinnaker does not natively integrate with all tools. In order for companies to adopt Spinnaker, they will need to add some functionality themselves. The way to add functionality now is to create extensions of the microservices. This is essentially treating a service like a library, making a new service, and importing the base service. To be successful, companies will need to manage building, testing, and distributing their extension. This is not a viable option for all companies due to the cognitive overhead and developer time needed to create the extensions.

### Lower Barrier of Entry to the Community

Diving right into an OSS community like Spinnaker can be daunting. It is a large project with a reasonably sized codebase and multiple services. There are a lot of people involved and several large companies as stakeholders. It takes a little while to get acclimated. Additionally, participants have to completely buy into the direction and methodologies of the project.

### Project Stability

It happens quite often that functionality added to the core OSS projects end up breaking other already existing functionality. With a plugin system, some of what might normally go into the core could be added as plugins instead. Creating isolation of functionality puts the responsibility of fixing problems with that functionality on the plugin owner. Furthermore, users that have no interest in the plugin won’t be impacted.

### Shareability

Being able to easily share functionality has been challenging for the OSS community. The options right now are to either get it into the upstream OSS project or create a custom distribution. Both options come with their own overhead. A plugin system bypasses this. While running one of the official OSS releases, a plugin can simply be enabled.

### Productivity

With a well thought out plugin system, community members would have a launch point to add functionality. It wouldn’t be necessary to dig deep into the Spinnaker codebase to figure out how to add something.

### Requirements
- Spinnaker services do not have to be rebuilt/recompiled.
- Plugin can be shared/distributed.

### Implementation

To shed light on the level of difficulty to implement a simple plugin system here is one possible approach. This example adds a custom stage to Orca.

1. Create a package in Orca called orca-plugin with the following class:
```
package com.netflix.spinnaker.orca.plugin.config;

import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;

@Configuration
@ConditionalOnProperty(prefix = "plugins", value = "enabled", matchIfMissing = false)
@ComponentScan("com.netflix.spinnaker.orca.plugin.stage")
public class PluginConfiguration {
 public PluginConfiguration() {
   System.out.println("Plugins Enabled.");
 }
}
```

2. Create a new Java project and import orca-core as a library:
```
package com.netflix.spinnaker.orca.plugin.stage;

import com.netflix.spinnaker.orca.CancellableStage;
import com.netflix.spinnaker.orca.pipeline.StageDefinitionBuilder;
import com.netflix.spinnaker.orca.pipeline.TaskNode.Builder;
import com.netflix.spinnaker.orca.pipeline.model.Stage;
import org.springframework.stereotype.Component;

@Component
public class MyStage implements StageDefinitionBuilder, CancellableStage {
 MyStage() {
   System.out.println("MyStage Constructed.");
 }

 public void taskGraph(Stage stage, Builder builder) {}

 public Result cancel(Stage stage) {
   return null;
 }
}
```
3. Build MyStage into a jar.
4. Run Orca with MyStage.jar in the classpath.
5. Profit.

The example given is extremely simplified and is not meant to serve as a full technical specification. Once there is community alignment on the higher level goals, we can all dig into implementation details.

### UI in Deck

React components can be hot loaded with tools like: https://github.com/gaearon/react-hot-loader

## The Vision

Being able to package up individual plugins is really just the beginning. Once that basic functionality is in place a lot more can be done.

### Halyard

Since Halyard controls deploying Spinnaker, it can also control activating plugins. The workflow could be something like:
```
$ hal plugins enable
$ hal plugins add github.com/acme/mystage
$ hal plugins edit mystage --some-parameter some-value
$ hal deploy apply
```
In this workflow, Halyard fetches the plugin from GitHub. It then adds the plugin to the dependencies while staging. Finally, Halyard deploys the plugin along with other configuration assets.

### JDK

Making the plugin system extremely easy to use will be a key factor in its success. Other projects have released a JDK for major IDEs. One of the most helpful was the Android JDK. Making a similar tool, where developers can select what type of plugin they wanted to create and have the project bootstrapped for them, can create a very seamless user experience.

### Plugin Types

So far there have been requests for:
- Stages
- Clouddrivers
- Pipeline and stage lifecycle hooks

### Marketplace

A marketplace could be a central index with metadata about the available plugins. It can hold information around versioning, compatibility, license, etc. Other projects have had great success with marketplaces. Such as Docker, Chef, etc. 

The Spinnaker community has grown quite a bit and finding marketplace launch partners shouldn't be a problem.

### Other Extensibility Approaches

There are many other ways to extend Spinnaker: Run Docker container stage, improved webhook stage, etc. A combination of many different extensibility options can really help catapult the project forward. Not all use cases are equal.

**Comment by andrewbackes** [comment]
@spinnakerbot add-label proposal

**Comment by andrewbackes** [comment]
@spinnakerbot add-label no-lifecycle

**Comment by robzienert** [comment]
First, thanks for taking the initiative to get the conversation going! Some thoughts, no particular order. I'm sorry this is long / brain-dumpy. I love what you have so far! 😄 

I think it's correct to not focus on a Docker- or webhook-centric model for extensibility. These things can be added later, but would introduce a lot of additional, hard problems (would we need to build an SLA model for plugins? What happens if a plugin is unavailable? Do we ignore the error / deactivate / hard fail the system? How do we know that the plugin is critical path or not?). 

The focus on (as of today), internal interfaces is intriguing. We do this at Netflix today for all of our extensions, but raises some challenges. How do we handle minimum supported versions for plugins if people are able to just plug right into Spring's application context and bean lifecycle? Are all interfaces now public contracts? I think for a first pass, this gets us business value, but will sacrifice core maintainers' abilities to rapidly iterate what was previously "internal-only" code. With that in mind, have you seen [Spring Plugin](https://github.com/spring-projects/spring-plugin)? The POC I was working on used this.

I think we need to have a little more structure around plugins beyond just a `@Configuration` class that adds Beans into the mix. Something in kork (for example) that would allow plugin developers to create a Manifest, similar to other plugin systems, that include a name, description, source code permalink, contact information, tags and minimum service version requirements. 

Have you considered the ramifications of plugins sharing the same classloader as the core service? What about plugins including their own dependencies, what if those dependencies clash with core service dependencies? Maybe we just acknowledge that this is a problem and look towards a v2 plugin system that accounts for plugin classloaders and better sandboxing?

This is good progress, but I'd definitely like to see "Vision" fleshed out more. Where are we headed? What do we want the end capabilities (pie-in-the-sky) to be, how does the initial RFC deliverables move us forward towards that goal, and perhaps what the next steps may be to getting us to the end vision? I'd also be interested in seeing something about _why_ we're not going with an RPC plugin system for the sake of explicitness; is it something that's in the end vision? 🤷‍♂️ 

What about plugins that span multiple services? Is it one jar-per-service? How do we ensure that all plugins are installed to all the necessary services in order for things to work? 

I think it'd be handy to bake in some default metrics for plugins as well, probably, so people know how often a plugin is being invoked, its error rate and so-on. Furthermore, we should think about how to make maintenance of a plugin better. Can we think of ways that error reporting to the author could be made easier for operators of Spinnaker? If I install a plugin and it starts producing errors, it'd be neat to dump anonymized data about the deployment of Spinnaker that could be shipped to the plugin author if requested. I'm not sure that specific example is something we want to bite off, but I would like to see more thought put into the plugin iteration and maintenance lifecycle.

Also, what's the expectations of a plugin developer? Should we be providing a TCK for plugin developers to minimally test against (e.g. assert the plugin correctly registers, is on the correct component scan path, etc etc).

**Comment by andrewbackes** [comment]
@robzienert -

Thank you for all of the feedback! It's great to see so much excitement around this.

I've tried to categorize what you've brought up in order to make sure it all gets addressed. Here is my attempt =P


### Plugin SLA / Multiple Jars / etc.

The initial thought was to use Halyard to enable and configure plugins. This is similar to how other configuration assets (java key stores, credentials, etc.) are managed.
1. User specifies the addition of a plugin to their Spinnaker installation via Halyard.
2. Halyard downloads the plugin's assets (jar/tarbal/etc.) locally.
3. If the plugin is not available, then Halyard fails and does not add the plugin to the configuration.
4. Halyard validates the plugin. The first version of the plugin system should at least check that the plugin is compatible with the version of Spinnaker selected.
5. User redeploys Spinnaker via Halyard.
6. Halyard delivers the plugin assets to the deployment target (in Kubernetes this is a secret/configmap).
7. Halyard wires up the correct configuration values in the service yaml files to use the delivered asset.
8. Halyard continues in its normal fashion to finish deploying Spinnaker.

Halyard would essentially treat plugins the same way it does configuration files. If a plugin causes a failure in one of the services then Halyard reports the failure at deploy time. At this point, determining how to proceed would be left to the user.

Note: Through this flow, it is expected that Halyard makes sure the plugin assets get to the correct Spinnaker services.

### Internal Interfaces & Public Contracts

One idea is to create predetermined hook points. Only these interfaces would be public contracts. This will limit the types of things plugins can do, but the number of hooks can grow over time. We could limit the first version to only have hooks for creating stages, for instance. This approach would essentially be taking the stance that core maintainers need to be prioritized over plugin creators.

### Sandboxing

Your suggestion to push dependency collision management off to a future version is a great idea! haha

## Deliverables

The original post mashed motivations, deliverables, and implementation together into one conversation. It is probably a better approach to start by getting consensus on deliverables. Once we are all aligned there, then we can dive into implementation options.


### Milestone 1

Goal: Lay the initial groundwork of a plugin system.
Success Criteria: A developer experienced with the Orca and Deck codebases can create and distribute a plugin.
Scope:
- Hooks for adding a stage to Orca
- Hooks for showing stage UI in Deck
- Functionality in Halyard for managing plugins

### Milestone 2

Goal: Help plugin creators get started.
Success Criteria: A developer who is not intimately experienced with the Spinnaker codebase can create and distribute a plugin.
Scope:
- A template project for Plugin creators to use when creating stages
- Documentation around the hooks in Orca and Deck
- A guide on how to create a plugin stage using the template project

### Milestone 3

Goal: Create a plugin marketplace.
Success Criteria: Operators of Spinnaker can find available plugins. Plugin creators can list their plugins. 
Scope:
- Website/Webapp listing available plugins (and any metadata/compatibility information)
- Plugin creators can share what they have created
- Spinnaker operators can find plugins


### Future (Pie in the Sky?)

The bigger vision is to have the plugin marketplace be the hub for the 3rd party Spinnaker ecosystem. Directionally, plugin creators and Spinnaker operators would be the customers and their experience should be optimized.

Once the baseline features are in place, we can start getting feedback. But naively, it feels like some of the following might be in scope:
- Pipeline/stage lifecycle hooks
- Infrastructure hooks
- Visualization hooks
- Developer SDK: Something more robust than a template project. A Sandbox for creating plugins without the need of standing up an entire Spinnaker stack.
- Usage data back to plugin developers
- More elaborate documentation.
- Marketplace automated compatibility testing
- Marketplace security vulnerability scanning

## General Extensibility

What is proposed above should not limit other extensibility approaches. It is meant to be lightweight and limited. In fact, this approach might just be a stepping stone towards a more elaborate system. Other approaches such (like RPC) should still be possible. We can probably run feature experiments with all of these approaches in parallel and let the market decide.


Let me know if you want additional details on the direction, scope, or deliverables. The hope is to get consensus and start digging with different technical approaches as soon as possible.

**Comment by skim1420** [comment]
I propose we look at [Task CRD](https://github.com/tektoncd/pipeline/blob/master/docs/tasks.md) as an extensibility/pluggability mechanism for Spinnaker custom stages. This doesn't work for everyone, I understand, so perhaps not the only extensibility mechanism. If it's better to do so, I can open a separate proposal issue for Task CRD.

**Comment by ethanfrogers** [comment]
@skim1420 the topic of Tekton has come up a few times this week. Having a proposal for that would be great so we could point folks to it as needed.

I agree that neither of these should be the only extension point. Tekton could provide a way for Kubernetes users to supplement their pipelines without writing any Java but Plugins could provide native feeling integrations (not just stages, perhaps data sources as well) without having to maintain a separate build.

**Comment by AbdulRahmanAlHamali** [comment]
:wave: I am just a user of Spinnaker, and not very experienced with it yet. But here is my two cents: Whenever I hear the word pluggable system, the first thing that comes to mind is Jenkins. While it is awesome how pluggable Jenkins is (sometimes I joke that you could literally turn it into anything you want, a MMO-RPG for example :joy:), this comes with a massive toll on the stability of the system, the compatibility of the plugins, etc.

I think that if Spinnaker wants to go the pluggable way. It should keep it very limited, very well-defined, and restrict it to only the places where pluggability produces the most value.

Thank you for the awesome work!

**Comment by rajeevb2** [comment]
@andrewbackes: What is the process to get traction on this?

**Comment by link108** [comment]
Hi @rajeevb2, this project has been moving forward - plugins are now a beta feature and are able to be used in Orca with the 1.16 Release. Let me know if you have any questions or use cases, I'd love to chat!

**Comment by plumpy** [comment]
This RFC is very old. The old RFC process has been superseded by a new one. If you still care about this RFC and would like feedback, please move it over to [the new RFC process](https://www.spinnaker.io/community/contributing/submitting/#before-you-begin). Thanks.
